test/test_selectors.py

```diff
@@ -0,0 +1,622 @@
+        self.assertEqual(list(keys.values()), [key])
+
+        # Unknown fileobj
+        self.assertRaises(KeyError, keys.__getitem__, 999999)
+
+        # Read-only mapping
+        with self.assertRaises(TypeError):
+            del keys[rd]
+
+    def test_register(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        # Ensure that the file is not yet added.
+        self.assertEqual(0, len(s.get_map()))
+        self.assertRaises(KeyError, lambda: s.get_map()[rd.fileno()])
+        self.assertRaises(KeyError, s.get_key, rd)
+        self.assertEqual(None, s._key_from_fd(rd.fileno()))
+
+        key = s.register(rd, selectors.EVENT_READ, "data")
+        self.assertIsInstance(key, selectors.SelectorKey)
+        self.assertEqual(key.fileobj, rd)
+        self.assertEqual(key.fd, rd.fileno())
+        self.assertEqual(key.events, selectors.EVENT_READ)
+        self.assertEqual(key.data, "data")
+        self.assertEqual(1, len(s.get_map()))
+        for fd in s.get_map():
+            self.assertEqual(fd, rd.fileno())
+
+    def test_register_bad_event(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        self.assertRaises(ValueError, s.register, rd, 99999)
+
+    def test_register_negative_fd(self):
+        s = self.make_selector()
+        self.assertRaises(ValueError, s.register, -1, selectors.EVENT_READ)
+
+    def test_register_invalid_fileobj(self):
+        s = self.make_selector()
+        self.assertRaises(ValueError, s.register, "string", selectors.EVENT_READ)
+
+    def test_reregister_fd_same_fileobj(self):
+        s, rd, wr = self.standard_setup()
+        self.assertRaises(KeyError, s.register, rd, selectors.EVENT_READ)
+
+    def test_reregister_fd_different_fileobj(self):
+        s, rd, wr = self.standard_setup()
+        self.assertRaises(KeyError, s.register, rd.fileno(), selectors.EVENT_READ)
+
+    def test_context_manager(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        with s as sel:
+            sel.register(rd, selectors.EVENT_READ)
+            sel.register(wr, selectors.EVENT_WRITE)
+
+        self.assertRaises(RuntimeError, s.get_key, rd)
+        self.assertRaises(RuntimeError, s.get_key, wr)
+
+    def test_unregister(self):
+        s, rd, wr = self.standard_setup()
+        s.unregister(rd)
+
+        self.assertRaises(KeyError, s.unregister, 99999)
+
+    def test_reunregister(self):
+        s, rd, wr = self.standard_setup()
+        s.unregister(rd)
+
+        self.assertRaises(KeyError, s.unregister, rd)
+
+    def test_unregister_after_fd_close(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        rdfd = rd.fileno()
+        wrfd = wr.fileno()
+        s.register(rdfd, selectors.EVENT_READ)
+        s.register(wrfd, selectors.EVENT_WRITE)
+
+        rd.close()
+        wr.close()
+
+        s.unregister(rdfd)
+        s.unregister(wrfd)
```

Comment:
This test and the one below it don't actually make any assertions. They should be refactored so that they do.